File: web/lib/onboarding.ts
/**
 * The five onboarding questions.
 *
 * Each step carries its own "why" note, which the design pins beside the
 * question on a sticky note. That note is the whole reason the flow feels
 * unlike a signup form: every question says what it will be used for before it
 * asks, which is also what §02 of the blueprint requires — ask only for what is
 * genuinely needed.
 */

export type StepKind =
  | "date"
  | "birthdate"
  | "choice"
  | "multi"
  | "place"
  | "contact"
  | "supporting";

export interface Step {
  kind: StepKind;
  eyebrow: string;
  question: string;
  help: string;
  why: string;
  noteBg: string;
  noteFg: string;
  skip?: string;
  options?: { label: string; sub?: string }[];
  /**
   * For a "choice" step, which answer it writes.
   *
   * Its absence is what marks the contact-window step: that one is a clock and
   * an "any time" toggle rather than a list, so it has options to render but
   * no single answer field to bind to.
   */
  field?: "feeding" | "contactRelationship";
}

/** Someone expecting. The original five. */
export const EXPECTING_STEPS: Step[] = [
  {
    kind: "date",
    eyebrow: "Step one",
    question: "When is your baby due?",
    help: "This anchors everything else — which week you are in, what is coming up, and when a test or visit is getting late.",
    why: "Nnneva asks once, and then never again. Every reminder, every deadline and every piece of preparation is worked out from this one date.",
    noteBg: "#FDF1F5",
    noteFg: "#0B2C22",
    skip: "I am not sure of my due date",
  },
  {
    kind: "place",
    eyebrow: "Step two",
    question: "Where will you have your care?",
    help: "So appointments, travel time and preparation are planned around the right place.",
    why: "Knowing the clinic means Nnneva can judge how long the journey takes and remind you to leave in time, rather than reminding you when you are already late.",
    noteBg: "#E4F1EB",
    noteFg: "#0B2C22",
    skip: "I will decide later",
    options: [
      { label: "An antenatal clinic", sub: "Standalone or attached to a hospital" },
      { label: "A hospital", sub: "Public or private" },
      { label: "A private midwife", sub: "Home or clinic visits" },
      { label: "Not decided yet", sub: "Nnneva will help you choose" },
    ],
  },
  {
    kind: "multi",
    eyebrow: "Step three",
    question: "What should Nnneva take off your hands?",
    help: "Pick as many as you like. You can change this at any time.",
    why: "This decides what Nnneva does on its own and what it leaves alone. Anything you do not pick here, it will not go near.",
    noteBg: "#FDF1F5",
    noteFg: "#0B2C22",
    options: [
      { label: "Appointments and preparation", sub: "Questions for the midwife, what to bring" },
      { label: "Tests and results", sub: "Booking, fasting, chasing results" },
      { label: "Everyday errands", sub: "Prescriptions, transport, supplies" },
      { label: "Preparing for the birth", sub: "Hospital bag, plans, contacts" },
    ],
  },
  {
    kind: "contact",
    eyebrow: "Step four",
    question: "Is there someone helping you?",
    help: "A partner, a family member, a friend. Optional, and they only ever see what you switch on.",
    why: "Nothing about your pregnancy reaches a trusted contact unless you approve it, and Nnneva asks every single time — not once at setup.",
    noteBg: "#E4F1EB",
    noteFg: "#0B2C22",
    skip: "It is just me for now",
    options: [],
  },
  {
    kind: "choice",
    eyebrow: "Step five",
    question: "When should Nnneva reach you?",
    help: "Nnneva only gets in touch when something needs a decision or a deadline is close.",
    why: "Everything that is not urgent waits for the window you pick here. A reminder at a useless hour is one more thing to carry.",
    noteBg: "#FDF1F5",
    noteFg: "#0B2C22",
    // No options: this step is a clock plus an "any time" choice, so the answer
    // is a time rather than one of a fixed few. See ContactWindow.
  },
];

/** Someone who has given birth.
 *
 * Not the pregnancy flow with the due date swapped out. A due date is a
 * prediction and a birth date is a fact, the useful help is different, and
 * asking about trimesters after the event would be the clearest possible sign
 * that nobody thought about this person. */
export const POSTPARTUM_STEPS: Step[] = [
  {
    kind: "birthdate",
    eyebrow: "Step one",
    question: "When was your baby born?",
    help: "This anchors everything else — where you are in your recovery, and which checks are coming up.",
    why: "Nnneva asks once. Postnatal checks, vaccinations and reviews all fall at known points after the birth, so this one date is enough to keep track of them.",
    noteBg: "#FDF1F5",
    noteFg: "#0B2C22",
  },
  {
    kind: "choice",
    eyebrow: "Step two",
    question: "How is feeding going?",
    field: "feeding",
    help: "So advice and reminders match what you are actually doing.",
    why: "This changes what is worth reminding you about and what would be noise. It is never used to judge — there is no right answer here, and \u201cstill working it out\u201d is a real one.",
    noteBg: "#E4F1EB",
    noteFg: "#0B2C22",
    skip: "I would rather not say",
    options: [
      { label: "Breastfeeding", sub: "Exclusively or mostly" },
      { label: "Formula", sub: "Exclusively or mostly" },
      { label: "Both", sub: "Mixed feeding" },
      { label: "Still working it out", sub: "It is early days" },
    ],
  },
  {
    kind: "multi",
    eyebrow: "Step three",
    question: "What should Nnneva take off your hands?",
    help: "Pick as many as you like. You can change this at any time.",
    why: "This decides what Nnneva does on its own and what it leaves alone. Anything you do not pick here, it will not go near.",
    noteBg: "#FDF1F5",
    noteFg: "#0B2C22",
    options: [
      { label: "Your own recovery", sub: "Postnatal checks, bleeding, pain, mood" },
      { label: "The baby's appointments", sub: "Weigh-ins, vaccinations, reviews" },
      { label: "Feeding and sleep", sub: "Routines, night feeds, what to expect" },
      { label: "Everyday errands", sub: "Prescriptions, transport, supplies" },
    ],
  },
  {
    kind: "contact",
    eyebrow: "Step four",
    question: "Is there someone helping you?",
    help: "A partner, a family member, a friend. Optional, and they only ever see what you switch on.",
    why: "Nothing reaches a trusted contact unless you approve it, and Nnneva asks every single time — not once at setup.",
    noteBg: "#E4F1EB",
    noteFg: "#0B2C22",
    skip: "It is just me for now",
    options: [],
  },
  {
    kind: "choice",
    eyebrow: "Step five",
    question: "When should Nnneva reach you?",
    help: "Nnneva only gets in touch when something needs a decision or a deadline is close.",
    why: "Everything that is not urgent waits for the window you pick here. With a newborn, a reminder at the wrong hour is worse than no reminder at all.",
    noteBg: "#FDF1F5",
    noteFg: "#0B2C22",
  },
];

/** Someone supporting another person.
 *
 * Three questions, because there are only three things Nnneva needs from
 * them. They have no due date and no clinic — those belong to the person they
 * are helping, and asking here would be asking them to answer for someone
 * else. */
export const SUPPORTER_STEPS: Step[] = [
  {
    kind: "supporting",
    eyebrow: "Step one",
    question: "Who are you supporting?",
    help: "Their Nnneva username. They decide whether to accept, and what you can see.",
    why: "Nothing is shared by you asking. They get a request, and until they accept it and switch something on, you see nothing at all.",
    noteBg: "#FDF1F5",
    noteFg: "#0B2C22",
    skip: "They have not joined yet",
  },
  {
    kind: "choice",
    eyebrow: "Step two",
    question: "What are you to them?",
    field: "contactRelationship",
    help: "So they know who is asking when the request arrives.",
    why: "\u201cChidi wants to help as your partner\u201d is a clearer thing to accept than a name on its own.",
    noteBg: "#E4F1EB",
    noteFg: "#0B2C22",
    options: [
      { label: "Partner" },
      { label: "Husband" },
      { label: "Mother" },
      { label: "Mother-in-law" },
      { label: "Sister" },
      { label: "Friend" },
      { label: "Doula" },
      { label: "Other family" },
    ],
  },
  {
    kind: "choice",
    eyebrow: "Step three",
    question: "When should Nnneva reach you?",
    help: "Only when something has actually been asked of you.",
    why: "You will hear from Nnneva when they hand you something to do, and not otherwise.",
    noteBg: "#FDF1F5",
    noteFg: "#0B2C22",
  },
];

export type Role = "expecting" | "postpartum" | "supporter";

export function stepsFor(role: Role | null): Step[] {
  if (role === "postpartum") return POSTPARTUM_STEPS;
  if (role === "supporter") return SUPPORTER_STEPS;
  return EXPECTING_STEPS;
}

/**
 * The help areas, taken from the steps that ask for them rather than repeated.
 * The profile screen offers the same set, and a second hand-written list would
 * drift — a label edited in one place would silently stop matching what is
 * already stored against accounts set up through the other.
 *
 * Both flows contribute, because the profile screen has to be able to render
 * whichever set an account was actually onboarded with.
 */
export const HELP_AREAS: string[] = [
  ...(EXPECTING_STEPS.find((s) => s.kind === "multi")?.options?.map((o) => o.label) ?? []),
  ...(POSTPARTUM_STEPS.find((s) => s.kind === "multi")?.options?.map((o) => o.label) ?? []),
];

/** Kept so existing imports of the pregnancy flow keep meaning what they did. */
export const STEPS = EXPECTING_STEPS;

File: api/app/routers/onboarding.py
"""Onboarding, which is a different set of questions for each kind of person.

Three roles, three flows, and deliberately no shared "and also" questions that
only apply to one of them. Someone supporting a friend has no due date and no
clinic; asking anyway and then treating the blanks as missing data is how a
product ends up nagging people for things they cannot give it.

What the three do have in common is the shape: ask only for what is needed,
say what each answer is for, and seed memory from the answers so the agent's
first run already has context (§02, §06).
"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, status
from sqlalchemy import select

from app.deps import CurrentUser, DbSession
from app.models import (
    ContactRequest,
    Memory,
    MemoryKind,
    PregnancyProfile,
    RequestKind,
    RequestStatus,
    TrustedContact,
    User,
    UserRole,
)
from app.schemas import OnboardingIn, ProfileOut
from app.routers.profile import profile_payload

router = APIRouter(prefix="/onboarding", tags=["onboarding"])


@router.post("", response_model=ProfileOut, status_code=status.HTTP_201_CREATED)
def complete_onboarding(payload: OnboardingIn, user: CurrentUser, db: DbSession) -> ProfileOut:
    # An account that reaches here without having answered the first-run
    # question is treated as expecting, which is what it was before the
    # question existed and what the migration assumed for older rows.
    role = user.role or UserRole.expecting
    if user.role is None:
        user.role = role

    if role is UserRole.supporter:
        _supporter(payload, user, db)
    else:
        _mother(payload, user, db, role)

    if payload.contact_window:
        user.contact_window = payload.contact_window

    db.commit()
    db.refresh(user)
    return profile_payload(user)


# ---- Someone expecting, or newly a parent ----------------------------------


def _mother(payload: OnboardingIn, user: User, db: DbSession, role: UserRole) -> None:
    """The care profile, anchored on whichever date the role implies."""
    expecting = role is UserRole.expecting

    if expecting and payload.due_date is None:
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_CONTENT, "A due date is needed to set this up."
        )
    if not expecting and payload.birth_date is None:
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_CONTENT,
            "A date of birth is needed to set this up.",
        )

    profile = user.profile or PregnancyProfile(user_id=user.id)
    # Only ever one of the two. Setting both would leave the week count
    # ambiguous, and clearing the other is what makes "I have given birth"
    # something an existing account could later say.
    profile.due_date = payload.due_date if expecting else None
    profile.birth_date = None if expecting else payload.birth_date
    profile.feeding = None if expecting else payload.feeding
    profile.care_location = payload.care_location
    profile.clinician = payload.clinician
    profile.help_areas = "\n".join(a.strip() for a in payload.help_areas if a.strip())
    db.add(profile)

    if payload.contact_name:
        contact = user.contacts[0] if user.contacts else TrustedContact(user_id=user.id, name="")
        contact.name = payload.contact_name.strip()
        contact.relationship_label = payload.contact_relationship or "Partner"
        contact.phone = (payload.contact_phone or "").strip() or None
        contact.email = (payload.contact_email or "").strip().lower() or None
        contact.can_see_shared_tasks = payload.contact_can_see_shared_tasks
        db.add(contact)

    db.flush()

    if expecting:
        anchor = (
            f"Due {payload.due_date:%-d %B %Y}. Currently {profile.gestational_week} weeks."
        )
    else:
        anchor = (
            f"Baby born {payload.birth_date:%-d %B %Y}. "
            f"Currently {profile.postnatal_week} weeks postpartum."
        )
    facts = [(MemoryKind.context, anchor)]

    if not expecting and payload.feeding:
        facts.append((MemoryKind.context, f"Feeding: {payload.feeding.lower()}."))
    if payload.care_location:
        facts.append((MemoryKind.context, f"Care at {payload.care_location}."))
    if payload.contact_name:
        facts.append((
            MemoryKind.preference,
            f"{payload.contact_name} is the trusted contact, permissions limited.",
        ))
    _remember(db, user, payload, facts)


# ---- Someone helping another person ----------------------------------------


def _supporter(payload: OnboardingIn, user: User, db: DbSession) -> None:
    """No care profile at all — they are not the one being cared for.

    The one thing this flow can do is put them in touch. A request from this
    side runs the opposite way to the usual one, and is still answered by the
    person being supported, so nothing is shared until she says yes.
    """
    facts: list[tuple[MemoryKind, str]] = []

    handle = (payload.supporting_username or "").strip().lower()
    if handle:
        other = db.scalars(select(User).where(User.username == handle)).first()
        # 404 for "no such handle" and for their own, exactly as /people does:
        # telling them apart would make this a way to test whether a handle
        # exists without signing in as anyone in particular.
        if other is None or other.id == user.id:
            raise HTTPException(status.HTTP_404_NOT_FOUND, "No account with that username")

        already = db.scalars(
            select(ContactRequest).where(
                ContactRequest.requester_id == user.id,
                ContactRequest.addressee_id == other.id,
            )
        ).first()
        if already is None:
            db.add(
                ContactRequest(
                    requester_id=user.id,
                    addressee_id=other.id,
                    relationship_label=payload.contact_relationship or "Partner",
                    kind=RequestKind.to_help,
                )
            )
        elif already.status is not RequestStatus.pending:
            already.status = RequestStatus.pending
            already.kind = RequestKind.to_help
            already.responded_at = None

        facts.append((
            MemoryKind.preference,
            f"Supporting @{other.username} as their "
            f"{(payload.contact_relationship or 'partner').lower()}.",
        ))

    _remember(db, user, payload, facts)


def _remember(db: DbSession, user: User, payload: OnboardingIn, facts: list) -> None:
    """Seed memory so the agent's first run never asks for what was just typed."""
    if payload.contact_window:
        facts.append(
            (MemoryKind.preference, f"Prefers contact: {payload.contact_window.lower()}.")
        )
    for kind, fact in facts:
        db.add(Memory(user_id=user.id, kind=kind, fact=fact, source="From onboarding"))
